http: config options for adding custom request headers

rshriram opened the pull request:
This PR allows the user to insert additional request headers into the HTTP request before it is forwarded by Envoy to the upstream cluster.

There are three levels at which this can be done:
1. At http connection manager level (across virtual hosts)
2. At a single virtual host level (across routes)
3. On a per route basis

These roughly correspond to Nginx's proxy_set_headers directive that can be applied at http/server/location block granularities.

Note that headers cannot be overridden. If duplicate headers are provided in Connection Manager, Virtual host and the route, only the first occurrence of the header (i.e. in the connection manager level) is taken into account.

mattklein123 commented:
@ccaraman please do first pass review.

ccaraman commented on test/common/router/config_impl_test.cc:
```diff
@@ -140,6 +140,11 @@ TEST(RouteMatcherTest, TestRoutes) {
     "x-envoy-upstream-canary",
     "x-envoy-virtual-cluster"
+  ],
+
```
nit: new line

ccaraman replied:
nvm.

ccaraman commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -28,6 +28,12 @@ void ConnectionManagerUtility::mutateRequestHeaders(Http::HeaderMap& request_hea
                                                     const Router::Config& route_config,
                                                     Runtime::RandomGenerator& random,
                                                     Runtime::Loader& runtime) {
+  // Add user-specified headers to the request before scrubbing the request headers.
```
General question: Why do we want to add these before scrubbing the headers?

rshriram replied:
I was chatting with Matt in Gitter. I am going to rework this a bit. I initially thought we could enforce some sort of rules on what headers the user can/cannot add. But the enforcement logic was messy. So, the decision was to have the code be simple. If users add bad configs (e.g., override authority/connection headers, etc.), the behavior will be undefined. We will mention this in the docs.

ccaraman commented on test/common/router/config_impl_test.cc:
```diff
@@ -279,6 +284,23 @@ TEST(RouteMatcherTest, TestRoutes) {
 
+  // Request header manipulation testing
+  {
+    // At route_config level
```
nit: incomplete sentence(this line and the next)

ccaraman commented on test/common/router/config_impl_test.cc:
```diff
@@ -279,6 +284,23 @@ TEST(RouteMatcherTest, TestRoutes) {
                                                 Http::LowerCaseString("x-envoy-virtual-cluster")}),
               ContainerEq(config.responseHeadersToRemove()));
 
+  // Request header manipulation testing
```
nit: full stop at the end of the sentence.

ccaraman commented on test/common/router/config_impl_test.cc:
```diff
@@ -279,6 +284,23 @@ TEST(RouteMatcherTest, TestRoutes) {
+                     {Http::LowerCaseString("Connection"), "Keep-Alive"}})),
+                ContainerEq(config.requestHeadersToAdd()));
+
+    Http::TestHeaderMapImpl headers = genHeaders("api.lyft.com", "/host/rewrite/me", "GET");
```
finalizeRequestHeaders doesn't invoke the added code for `request_headers_to_add`. Please add a test to https://github.com/lyft/envoy/blob/master/test/common/http/conn_manager_utility_test.cc

rshriram commented:
fixes #539

mattklein123 commented on source/common/router/config_impl.cc:
```diff
@@ -154,6 +161,17 @@ bool RouteEntryImplBase::matchRoute(const Http::HeaderMap& headers, uint64_t ran
 const std::string& RouteEntryImplBase::clusterName() const { return cluster_name_; }
 
 void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+  // Add user-specified request headers from the vhost and then from the route.
```
I'm not super thrilled that we do one thing in the connection manager and two things here. I'm wondering if we should move at least adding request headers for the global route level here. (We can leave munging response headers in the connection manager). Thoughts?

rshriram commented:
I agree. I was looking for a way to get a handle on the global config
object so that I can add all three headers here. But I couldn't go beyond
the virtual host. If there is already a way to do so, let me know.
Otherwise, I can look into adding an interface accordingly.
On Tue, Mar 28, 2017 at 6:10 PM Matt Klein <notifications@github.com> wrote:

> *@mattklein123* commented on this pull request.
> ------------------------------
>
> In source/common/router/config_impl.cc
> <https://github.com/lyft/envoy/pull/630#discussion_r108549845>:
>
> > @@ -154,6 +161,17 @@ bool RouteEntryImplBase::matchRoute(const Http::HeaderMap& headers, uint64_t ran
>  const std::string& RouteEntryImplBase::clusterName() const { return cluster_name_; }
>
>  void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers) const {
> +  // Add user-specified request headers from the vhost and then from the route.
>
> I'm not super thrilled that we do one thing in the connection manager and
> two things here. I'm wondering if we should move at least adding request
> headers for the global route level here. (We can leave munging response
> headers in the connection manager). Thoughts?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lyft/envoy/pull/630#pullrequestreview-29588418>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AH0qd2xhMjV6WtSgINHkLtMP0pH5DWuXks5rqYVngaJpZM4MqhCS>
> .
>
-- 

~shriram

mattklein123 commented:
You don't need any interface changes as you are inside config_impl. Just have a reference chain that leads back to the ConfigImpl root (via virtual host impl).

rshriram commented:
@mattklein123 @ccaraman I have moved the request header addition functionality completely into Config Impl (and removed all public interfaces as well). The hash map semantics of Http::HeaderMap are still unclear. It seems that the class refuses to overwrite headers that are already set -- this causes the test to fail. Any pointers?

ccaraman commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -338,6 +342,28 @@ header_name
+Adding custom request headers
+-----------------------------
+
+Custom request headers can be added to a request that matches a specific route. The headers are
```
Can you link to this from the other request_headers_to_add? That way you don't have to duplicate the logic.

ccaraman commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -338,6 +342,28 @@ header_name
+    {"key": "header2", "value": "value2"}
+  ]
+
+*Note:* In the presence of duplicate header keys, headers specified at the global
```
I don't agree with the ordering of precedence. The most specific rules (routes) should take precendence over the less specific rules virtual host and vh over  HTTP Conn Man route configurations.
The order of 
1. Route
2. Virtual Host
3. Route configuration within the HTTP Conn Man Filter. 

I believe this to be better because at the HTTP Conn Man layer we have request headers to be added for all requests then particular vhs or routes can set particular more specific header values.

ccaraman commented on docs/configuration/http_conn_man/route_config/route_config.rst:
```diff
@@ -47,6 +48,22 @@ response_headers_to_remove
+      {"key": "header2", "value": "value2"}
+    ]
+
+  *Note:* In the presence of duplicate header keys, only the first header is taken into
```
Please add link to documentation that mentions the order of precedence.

ccaraman commented on source/common/router/config_impl.cc:
```diff
@@ -533,8 +564,6 @@ VirtualHostImpl::virtualClusterFromEntries(const Http::HeaderMap& headers) const
 
 ConfigImpl::ConfigImpl(const Json::Object& config, Runtime::Loader& runtime,
                        Upstream::ClusterManager& cm, bool validate_clusters) {
-  route_matcher_.reset(new RouteMatcher(config, runtime, cm, validate_clusters));
```
does this need to be moved?

ccaraman commented on source/common/router/config_impl.cc:
```diff
@@ -154,6 +161,20 @@ bool RouteEntryImplBase::matchRoute(const Http::HeaderMap& headers, uint64_t ran
 void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+  // Add route-level headers first, followed by virtual host level headers
+  // and finally connection manager level headers If there are two headers with same
+  // name, we consider only the first one, and ignore the rest.
```
nit remove the comma before the and.

ccaraman commented on test/common/router/config_impl_test.cc:
```diff
@@ -337,6 +314,144 @@ TEST(RouteMatcherTest, TestRoutes) {
+          "prefix": "/",
+          "cluster": "instant-server",
+          "timeout_ms": 30000
+        }]
```
new line for `]`

rshriram replied:
done & done. Sorry, I rebased and it ended showing up as bunch of new commits.

mattklein123 commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -338,6 +342,27 @@ header_name
+  ]
+
+*Note:* In the presence of duplicate header keys, headers specified at the
+individual route-level take precedence over those specified at 
```
This is not exactly clear: For headers that are not in the O(1) set, the header *will* get appended multiple times. So there is precedent, but only in the ordering. This is kind of complicated, and not sure whether you want to try to describe this in the docs, or just rephrase to talk about the order of addition.

mattklein123 commented on source/common/router/config_impl.cc:
```diff
@@ -154,6 +161,20 @@ bool RouteEntryImplBase::matchRoute(const Http::HeaderMap& headers, uint64_t ran
 const std::string& RouteEntryImplBase::clusterName() const { return cluster_name_; }
 
 void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+  // For user-specified request headers, route-level headers of take precedence over
```
typo "of"

mattklein123 commented on source/common/router/config_impl.cc:
```diff
@@ -433,14 +463,15 @@ VirtualHostImpl::VirtualClusterEntry::VirtualClusterEntry(const Json::Object& vi
-    VirtualHostSharedPtr virtual_host(
-        new VirtualHostImpl(*virtual_host_config, runtime, cm, validate_clusters));
+  for (const Json::ObjectPtr& virtual_host_config : json_config.getObjectArray("virtual_hosts")) {
+    VirtualHostSharedPtr virtual_host(new VirtualHostImpl(*virtual_host_config, global_http_config,
```
s/global_http_config/global_route_config

mattklein123 commented on source/common/router/config_impl.h:
```diff
@@ -50,18 +50,23 @@ class SslRedirectRoute : public Route {
+  const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd() const {
+    return request_headers_to_add_;
+  }
+  const ConfigImpl& globalHttpConfig() const { return global_http_config_; }
```
globalRouteConfig() (just rename this similarly everywhere)

rshriram replied:
done

mattklein123 commented:
This change LGTM. I will let @ccaraman give you final review/approve/merge.
